src/passes/reconstruct-stringify-walker.cpp

```diff
@@ -0,0 +1,193 @@
+  // and have already created the body of the outlined function that will be
+  // called, then we will skip instructions, otherwise we add the instructions
+  // to the outlined function. If we are not in a sequence, then the
+  // instructions are sent to the existing function.
+  if (seqCounter < sequences.size() &&
+      instrCounter >= sequences[seqCounter].startIdx &&
+      instrCounter <= sequences[seqCounter].endIdx) {
+    return getModule()->getFunction(sequences[seqCounter].func)->body
+             ? InSkipSeq
+             : InSeq;
+  }
+  return NotInSeq;
+}
+
+void ReconstructStringifyWalker::maybeBeginSeq() {
+  auto currState = getCurrState();
+  if (currState != state) {
+    switch (currState) {
+      case NotInSeq:
+        return;
+      case InSeq:
+        transitionToInSeq();
+        break;
+      case InSkipSeq:
+        transitionToInSkipSeq();
+        break;
+    }
+  }
+  state = currState;
+}
+
+void ReconstructStringifyWalker::transitionToInSeq() {
+  Function* outlinedFunc = getModule()->getFunction(sequences[seqCounter].func);
+  if (auto val = outlinedBuilder.visitFunctionStart(outlinedFunc);
+      auto err = val.getErr()) {
+    Fatal() << err->msg;
+  }
+  // Add a local.get instruction for every parameter of the outlined function
+  Signature sig = outlinedFunc->type.getSignature();
+  for (Index i = 0; i < sig.params.size(); i++) {
+    if (auto val = outlinedBuilder.makeLocalGet(i); auto err = val.getErr()) {
+      Fatal() << err->msg;
+    }
+  }
+  // Make a call from the existing function to the outlined function. This call
+  // will replace the instructions moved to the outlined function
+  if (auto val = existingBuilder.makeCall(outlinedFunc->name, false);
+      auto err = val.getErr()) {
+    Fatal() << err->msg;
+  }
+  DBG(std::cout << "\ncreated outlined fn: " << outlinedFunc->name);
+}
+
+void ReconstructStringifyWalker::transitionToInSkipSeq() {
+  Function* outlinedFunc = getModule()->getFunction(sequences[seqCounter].func);
+  if (auto val = existingBuilder.makeCall(outlinedFunc->name, false);
+      auto err = val.getErr()) {
+    Fatal() << err->msg;
+  }
+  DBG(std::cout << "\n\nstarting to skip instructions "
+                << sequences[seqCounter].startIdx << " - "
+                << sequences[seqCounter].endIdx - 1 << " to "
+                << sequences[seqCounter].func << " and adding call() instead");
+}
+
+void ReconstructStringifyWalker::maybeEndSeq() {
+  if (instrCounter + 1 == sequences[seqCounter].endIdx) {
+    transitionToNotInSeq();
+  }
+}
+
+void ReconstructStringifyWalker::transitionToNotInSeq() {
+  if (state == InSeq) {
+    if (auto val = outlinedBuilder.visitEnd(); auto err = val.getErr()) {
+      Fatal() << err->msg;
+    }
+    DBG(std::cout << "\n\nEnd of sequence to " << &outlinedBuilder);
+  }
+  // Completed a sequence so increase the seqCounter and reset the state
+  seqCounter++;
+  state = NotInSeq;
```

Comment:
The other `transition` functions did not set the `state`. It would be good to be consistent about who is responsible for setting the `state`.

Reply:
Good call, moved to maybeEndSeq() since maybeBeginSeq() also sets state.